Bottom: 3d24ec64bc4fb2791200ef323eb44f376e225158
Top:    8bef0db9d47cd81fd48e402fff240633f0212c20
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-09-05 23:18:49 +0100

Refresh of hid-gtk-connect-up-the-file-ch

---

diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index 3e253cb..351a699 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -2169,6 +2169,8 @@ hid_gtk_init ()
   ghid_hid.flush_debug_draw         = ghid_flush_debug_draw;
   ghid_hid.finish_debug_draw        = ghid_finish_debug_draw;
 
+  ghid_hid.notify_save_pcb          = ghid_notify_save_pcb;
+
   hid_register_hid (&ghid_hid);
 #include "gtk_lists.h"
 }
diff --git a/src/hid/gtk/gui-top-window.c b/src/hid/gtk/gui-top-window.c
index 0b57540..f020043 100644
--- a/src/hid/gtk/gui-top-window.c
+++ b/src/hid/gtk/gui-top-window.c
@@ -346,10 +346,8 @@ file_changed_cb (GFileMonitor     *monitor,
 }
 
 static void
-connect_file_change_monitor (GhidGui *_gui)
+disconnect_file_change_monitor (GhidGui *_gui)
 {
-  GFile *file;
-
   if (_gui->file_monitor != NULL)
     g_object_unref (_gui->file_monitor);
   _gui->file_monitor = NULL;
@@ -357,6 +355,15 @@ connect_file_change_monitor (GhidGui *_gui)
   if (_gui->info_bar != NULL)
     gtk_widget_destroy (_gui->info_bar);
   _gui->info_bar = NULL;
+}
+
+static void
+connect_file_change_monitor (GhidGui *_gui)
+{
+  GFile *file;
+
+  /* Ensure any existing monitor is disconnected */
+  disconnect_file_change_monitor (_gui);
 
   if (PCB->Filename == NULL ||
       *PCB->Filename == '\0')
@@ -393,6 +400,15 @@ ghid_sync_with_new_layout (void)
   connect_file_change_monitor (ghidgui);
 }
 
+void
+ghid_notify_save_pcb (bool done)
+{
+  if (!done)
+    disconnect_file_change_monitor (ghidgui);
+  else
+    connect_file_change_monitor (ghidgui);
+}
+
 /* ---------------------------------------------------------------------------
  *
  * layer_process()
diff --git a/src/hid/gtk/gui.h b/src/hid/gtk/gui.h
index dfed2c6..95e302b 100644
--- a/src/hid/gtk/gui.h
+++ b/src/hid/gtk/gui.h
@@ -338,6 +338,7 @@ int ghid_drc_window_throw_dialog (void);
 
 /* In gui-top-window.c  */
 void ghid_update_toggle_flags (void);
+void ghid_notify_save_pcb (bool done);
 void ghid_install_accel_groups (GtkWindow *window, GhidGui *gui);
 void ghid_remove_accel_groups (GtkWindow *window, GhidGui *gui);
